soaktest/soakclient_transaction/soakclient_transaction.go

```diff
@@ -0,0 +1,552 @@
+package main
+
+/**
+ * Copyright 2021 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"github.com/confluentinc/confluent-kafka-go/kafka"
+	"os"
+	"os/signal"
+	"soaktest"
+	"sync"
+	"syscall"
+	"time"
+)
+
+const PRODUCER_TRANSACTION_COMMIT_SUCCEED = "producer.transaction.commit.succeed"
+const PRODUCER_TRANSACTION_COMMIT_FAILED = "producer.transaction.commit.failed"
+const PRODUCER_TRANSACTION_ABORT_SUCCEED = "producer.transaction.abort.succeed"
+const PRODUCER_TRANSACTION_ABORT_FAILED = "producer.transaction.abort.failed"
+const TRANSACTION = "transaction."
+
+var msgMissCntForInput uint64
+var msgDupCntForInput uint64
+var RETRY_NUM = 3
+var SLEEP_TIME = 3 * time.Second
+
+func main() {
+
+	broker := flag.String("broker", "127.0.0.1:9092", "bootstrap servers")
+	inputTopic := flag.String("inputTopic", "inputTopic",
+		"producer will deliver messages to this topic")
+	outTopic := flag.String("outTopic", "outTopic",
+		"consumer will consume messages from this topic")
+	groupId := flag.String("groupId", "groupId",
+		"the group consumer will join")
+	transactionId := flag.String("transactionId", "transactionId", "transaction id")
+	ccloudAPIKey := flag.String("ccloudAPIKey", "", "sasl username")
+	ccloudAPISecret := flag.String("ccloudAPISecret", "", "sasl password")
+
+	flag.Parse()
+
+	num, version := kafka.LibraryVersion()
+	fmt.Printf("num = %d, librdkafka %q\n", num, version)
+
+	var wg sync.WaitGroup
+	doneChan := make(chan bool)
+	errorChan := make(chan error)
+
+	wg.Add(1)
+	go soaktest.GetRusageMetrics(TRANSACTION, &wg, doneChan, errorChan)
+
+	wg.Add(1)
+	go verifyProducerTransaction(*broker, *inputTopic, *outTopic, *groupId,
+		*transactionId, *ccloudAPIKey, *ccloudAPISecret, &wg, doneChan, errorChan)
+
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)
+
+	select {
+	case sig := <-sigChan:
+		fmt.Printf("Signal caught: %v", sig)
+	case err := <-errorChan:
+		fmt.Printf("Error caught: %v", err)
+	}
+
```

Comment:
Same deadlock here, unbuffered channels will block on write until there is a reader.